File: Tarefa44/exercicio3.java
/*
 * Calcule a complexidade de tempo das soluções da série de Fibonacci
apresentadas no módulo e compare as duas.

Recursiva pura:       Tempo:O(2^n)	        Espaço:O(n)
Recursiva dinamica:   Tempo:O(n)	        Espaço:O(n)

A implementação recursiva pura tem complexidade exponencial, pois 
recalcula os mesmos valores várias vezes.
já a implementação dinâmica armazena os resultados intermediários, 
reduzindo a complexidade, o que a torna muito mais eficiente para 
valores maiores é muito mais eficiente e prática para números grandes.
 * 
 */

import java.math.BigInteger;

public class exercicio3 {

    public static int fibonacciPuro(int n) {
        if (n <= 1) {
            return n;
        }
        return fibonacciPuro(n - 1) + fibonacciPuro(n - 2);
    }

    // necessário usar BigInteger para numeros grandes pois int não suporta.
    public static BigInteger fibDinamico(int n, BigInteger[] memo) {
        if (memo[n] != null)
            return memo[n];
        if (n < 2)
            memo[n] = BigInteger.valueOf(n);
        else
            memo[n] = fibDinamico(n - 1, memo).add(fibDinamico(n - 2, memo));
        return memo[n];
    }

    public static void main(String[] args) {

        System.out.println("Fibonacci com Recursividade pura (12): " + fibonacciPuro(12));
        // System.out.println("Fibonacci com Recursividade pura (100): " + fibonacciPuro(100)); // Causará erro de Recursão Máxima

        int n = 100;
        BigInteger[] memo = new BigInteger[n + 1];
        System.out.println("Fibonacci com Programação Dinâmica (" + n + "): " + fibDinamico(n, memo));
    }
}

File: Tarefa44 Python/exercicio3.py
# calcule a complexidade de tempo das soluções da série de Fibonacci
# apresentadas no módulo e compare as duas.

# Implementação recursiva pura
def fibo(n):
    if n <= 1:
        return n;
    return fibo(n-1) + fibo(n-2);

print('RECURSIVIDADE PURA')
print('12ª posição do array = 13ª da sequencia de Fibonacci =', fibo(12));

print('30ª posição do array = 31ª da sequencia de Fibonacci =', fibo(30));

# Descomente a linha a baixo para ver o erro "Maximum Recursion Depth Exceeded"
#print('1000ª posição do array = 1001ª da sequencia de Fibonacci =', fibo(1000));

print(' ');

# Implementação dinâmica
memo_fib = {};

def fibonacci_dp(n):
    if n in memo_fib:
        return memo_fib[n];
    if n == 0 or n == 1:
        memo_fib[n] = n;
    else:
        memo_fib[n] = fibonacci_dp(n-1) + fibonacci_dp(n-2);
    return memo_fib[n];

print('RECURSIVIDADE DINAMICA')

print('12ª posição do array = 13ª da sequencia de Fibonacci =', fibonacci_dp(12));

print('30ª posição do array = 31ª da sequencia de Fibonacci =', fibonacci_dp(30));

print('1000ª posição do array = 1001ª da sequencia de Fibonacci =', fibonacci_dp(1000))


# Recursiva pura:       Tempo:O(2^n)	    Espaço:O(n)
# Recursiva dinamica:   Tempo:O(n)	        Espaço:O(n)

# A implementação recursiva pura tem complexidade exponencial, pois 
# recalcula os mesmos valores várias vezes.
# já a implementação dinâmica armazena os resultados intermediários, 
# reduzindo a complexidade, o que a torna muito mais eficiente para 
# valores maiores é muito mais eficiente e prática para números grandes.